update mutationObserver with debounce and detection of multiple classes

luckman212 opened the pull request:
Companion fix for https://github.com/JonasDoesThings/obsidian-hide-folders/issues/11 (this seems better than previous proposed fix from #15)

Adds a 100ms delay around mutationObserver, and changes the detection class from `nav-folder` to `nav-files-container` which seems to be correct (on my system, `nav-folder` was not present)

JonasDoesThings commented:
Hi, sorry for my slow reactions - it's great that you were able to further look into the whole issue :)
Weird that `nav-folder` didn't exist for you, for me every item in the file-tree has the class.

Anyways; is the `setTimeout` really also needed, or would it be enough to just change the detection class from `nav-folder` to `nav-files-container` for it to work on your system?

JonasDoesThings commented:
Oh and I noticed that we still need to check for both classes. In my setup when having collapsed folders with sub-folders that should get hidden, the sub-folders won't be hidden after expanding the parent-folder.

This is due to the mutationOberver's if-condition checking for changes of `nav-folder` items being removed by this PR.
We should add it back with an "or" in the if-statement, and just check if the mutationRecord contains one of the two classes.

JonasDoesThings commented on main.ts:
```diff
@@ -159,8 +159,10 @@ export default class HideFoldersPlugin extends Plugin {
       mutationRecord.forEach(record => {
-        if(record.target?.parentElement?.classList.contains("nav-folder")) {
-          this.processFolders();
+        if(record.target?.parentElement?.classList.contains("nav-files-container")) {
```
See my comment in the PR-thread, add back the old check and just check if the mutationRecord's parent classList contains either `nav-folder` or `nav-files-container`

JonasDoesThings reviewed (changes requested):
block merge for now

luckman212 commented:
@JonasDoesThings I just squashed and pushed an updated fix here.

I tested your theory about not needing the `setTimeout()` wrapper inside the MutationObserver, and you were right—it wasn't necessary. So I removed that.

I also reworked the `forEach()` so it operates on multiple classes.

Lastly, I noticed that `processFolders()` was getting called once for every file/folder added to my File Explorer on startup (6,000+ times in my vault!) which seemed very wasteful, so I added a debouncer to that function.

JonasDoesThings commented:
haha 6000 times is really not good :D nice find!
didn't even know obsidian has a built-in `debounce` helper, cool.

problem with that was that now it introduced a visible 500ms lag even for opening folders in small vaults (see video):
https://github.com/JonasDoesThings/obsidian-hide-folders/assets/5180118/c7bd3f41-d4fa-47ec-8c45-c9075f4f5ed7

Thus I have looked into other ways to lessening the calls to `processFolders`. I noticed that we loop through the MutationObserver's records. Opening a folder with 6-sub-folders would thus result in 6 MutationObserver mutationRecords which we would all loop through and for each one call `processFolders` if the classes matched. 

This is not necessary, if we instead again utilize `.some` and only check if ANY of the objects matches our class-check. After we found 1 object that fits this flter, we can call `processFolders` and do not need to look at the other mutationRecords.

With that change we only call `processFolders` 1 time instead of 6 times in the aforementioned scenario of expanding a single folder with 6 sub-folders.

**In my test-vault I could get the `processFolders` calls down from 47 times to only 5 times on startup!**

@luckman212 Could you checkout my branch ([`test/pr-16-reduce-callcount`](https://github.com/JonasDoesThings/obsidian-hide-folders/tree/test/pr-16-reduce-callcount)) locally on your end, install it, and check in the console of obsidian if the call-count goes down to a reasonable amount for you too?

If this already reduces the amount by a lot we can debounce-less and use a shorter delay :)
Hope my answer was not written too confusing 😅 currently working on many things in parallel

luckman212 commented:
@JonasDoesThings 🚀 This is awesome stuff! I didn't realize the debouncer was causing a delay like that. I expected the function to execute at the start of the delay, but I see from your video and my own testing that it executes at the _end_ which causes that annoying delay. Oops.

I tested your `test/pr-16-reduce-callcount` branch and indeed it worked very well. I just pushed a new squashed commit here.

I brought back the debounce, but reduced the delay from 500ms to 10ms. In my testing, this did not produce any noticeable difference in the file explorer tree updating, but did reduce the # of `processFolders()` calls from about 70 on my system down to 30 or so. So this seems to be the "best of both worlds". 

Curious if your testing confirms this, or if you find any more issues. Hopefully this is almost ready to merge.

JonasDoesThings commented:
Hi yeah, looks like a good compromise :) 
I will play around with it a bit and if everything seems to work I can release a new version with this PR later today! ❤️

luckman212 commented:
Just a note: I installed the new 1.6.5 Obsidian that has Electron 30 embedded. Patch still working fine so far.

JonasDoesThings commented:
Thx for all your work! I have released v1.5.1 now including your PR 🙏 

And thanks again for all your patience with my back and forth, but I think we were able to really find a good solution here :) 
Cheers from Austria

luckman212 commented:
Cheers, much appreciate your help here as well
